src/cast_core/src/CollectionScanner.cpp

```diff
@@ -34,31 +31,127 @@
     metrics::Operation scanOperation;
+    metrics::Operation exceptionsCaught;
+    int64_t documents;
+    int64_t scanSizeBytes;
+    ScanType scanType;
+    bool queryCollectionList;
+    std::optional<mongocxx::options::transaction> transactionOptions;
 
     PhaseConfig(PhaseContext& context,
                 const CollectionScanner* actor,
                 const mongocxx::database& db,
                 int collectionCount,
-                int threads)
-        : skipFirstLoop{context["SkipFirstLoop"].maybe<bool>().value_or(false)},
-          scanOperation{context.operation("Scan", actor->id())} {
-        // This tracks which CollectionScanners we are out of all CollectionScanners. As opposed to
-        // ActorId which is the overall actorId in the entire genny workload.
-        // Distribute the collections among the actors.
-        for (const auto& collectionName :
-             distributeCollectionNames(collectionCount, threads, actor->_index)) {
-            collections.push_back(db[collectionName]);
+                int threads,
+                bool generateCollectionNames)
+        : database{db},
+          skipFirstLoop{context["SkipFirstLoop"].maybe<bool>().value_or(false)},
+          filterExpr{context["Filter"].maybe<DocumentGenerator>(context, actor->id())},
+          scanOperation{context.operation("Scan", actor->id())},
+          exceptionsCaught{context.operation("ExceptionsCaught", actor->id())},
+          documents{context["Documents"].maybe<IntegerSpec>().value_or(0)},
+          scanSizeBytes{context["ScanSizeBytes"].maybe<IntegerSpec>().value_or(0)} {
+        // Initialise scan type enum.
+        auto scanTypeString = context["ScanType"].to<std::string>();
+        // Ignore case
+        boost::algorithm::to_lower(scanTypeString);
+        if (scanTypeString == "count") {
+            scanType = Count;
+        } else if (scanTypeString == "snapshot") {
+            transactionOptions = mongocxx::options::transaction{};
+            auto readConcern = mongocxx::read_concern{};
+            readConcern.acknowledge_level(mongocxx::read_concern::level::k_majority);
+            transactionOptions->read_concern(readConcern);
+            scanType = Snapshot;
+        } else if (scanTypeString == "standard") {
+            scanType = Standard;
+        } else {
+            BOOST_THROW_EXCEPTION(InvalidConfigurationException("ScanType is invalid."));
+        }
+        if (generateCollectionNames) {
+            queryCollectionList = false;
+            BOOST_LOG_TRIVIAL(info) << " Generating collection names";
+            for (const auto& collectionName :
+                 distributeCollectionNames(collectionCount, threads, actor->_index)) {
+                collections.push_back(db[collectionName]);
+            }
+        } else {
+            queryCollectionList = true;
         }
     }
 };
 
+void collectionScan(genny::v1::ActorPhase<CollectionScanner::PhaseConfig>& config,
```

Comment:
hrm...having to spell out a `v1` type is definitely not something we wanted you to have to do. I *think* you can instead have this be a `CollectionScanner::PhasConfig&` and pass it by calling `collectionScan(*config, collections)`.

(Similar comment applies to other functions doing the same thing)

Reply:
Seems to blow things up?